pkg/loki/loki.go

```diff
@@ -20,50 +23,158 @@ func init() {
 }
 
 type Loki struct {
-	promtails []*promtail.Promtail
+	mut sync.Mutex
+
+	reg       prometheus.Registerer
+	l         log.Logger
+	instances map[string]*Instance
 }
 
 // New creates and starts Loki log collection.
 func New(reg prometheus.Registerer, c Config, l log.Logger) (*Loki, error) {
 	l = log.With(l, "component", "loki")
 
+	loki := &Loki{
+		instances: make(map[string]*Instance),
+		reg:       reg,
+		l:         log.With(l, "component", "loki"),
+	}
+	if err := loki.ApplyConfig(c); err != nil {
+		return nil, err
+	}
+	return loki, nil
+}
+
+// ApplyConfig updates Loki with a new Config.
+func (l *Loki) ApplyConfig(c Config) error {
+	l.mut.Lock()
+	defer l.mut.Unlock()
+
 	if c.PositionsDirectory != "" {
 		err := os.MkdirAll(c.PositionsDirectory, 0700)
 		if err != nil {
-			level.Warn(l).Log("msg", "failed to create the positions directory. logs may be unable to save their position", "path", c.PositionsDirectory, "err", err)
+			level.Warn(l.l).Log("msg", "failed to create the positions directory. logs may be unable to save their position", "path", c.PositionsDirectory, "err", err)
 		}
 	}
 
-	promtails := make([]*promtail.Promtail, 0, len(c.Configs))
+	newInstances := make(map[string]*Instance, len(c.Configs))
 
 	for _, ic := range c.Configs {
-		if len(ic.ClientConfigs) == 0 {
-			level.Info(l).Log("msg", "skipping creation of a promtail because no client_configs are present", "config", ic.Name)
+		// If an old instance existed, update it and move it to the new map.
+		if old, ok := l.instances[ic.Name]; ok {
+			err := old.ApplyConfig(ic)
+			if err != nil {
+				return err
+			}
+
+			delete(l.instances, ic.Name)
```

Comment:
Since the entire `l.instances` map is being swapped below, shouldn't we avoid modifying it here in case an `old.ApplyConfig(ic)` fails? Seems like that would leave the map in an inconsistent state, right?

Reply:
Good point, let me think about this

Reply:
Thanks! I changed the way we detect stale instances.